adapters/Vungle/VungleAdapter/GADMAdapterVungleRouter.m

```diff
@@ -126,67 +126,87 @@ - (BOOL)addDelegate:(nonnull id<GADMAdapterVungleDelegate>)delegate {
-        GADMAdapterVungleBannerRequest *bannerRequest = nil;
-        while (bannerRequest = [enumerator nextObject]) {
-          if ([bannerRequest.placementID isEqualToString:delegate.bannerRequest.placementID]) {
+        GADMAdapterVungleBannerRequest *vungleBannerRequest = nil;
+        while (vungleBannerRequest = [enumerator nextObject]) {
+          // There is already a banner delegate with same placementID
+          // but different uniquePubRequestID in _bannerDelegates.
+          if ([vungleBannerRequest.placementID isEqualToString:delegate.bannerRequest.placementID]) {
             return NO;
-          } else {
-            NSOrderedSet *set = [_bannerDelegates objectForKey:bannerRequest];
-            for(id<GADMAdapterVungleDelegate> delegate in set) {
-              if (delegate.bannerState == BannerRouterDelegateStateWillPlay || delegate.bannerState == BannerRouterDelegateStatePlaying) {
-                return NO;
-              }
+          } else if (vungleBannerRequest.placementID.length > 0 && ![vungleBannerRequest.placementID isEqualToString:delegate.bannerRequest.placementID]) {
+            // There is already a banner delegate with different placementID in _bannerDelegates.
+            if (!_bannerRequest) {
+              _bannerRequest = [vungleBannerRequest copy];
             }
+            return NO;
           }
         }
 
-        NSMutableOrderedSet *set = [[NSMutableOrderedSet alloc] initWithObject:delegate];
-        GADMAdapterVungleMapTableSetObjectForKey(_bannerDelegates, delegate.bannerRequest, set);
-        _bannerPlacementID = [delegate.desiredPlacement copy];
-      } else if (delegate && [_bannerDelegates objectForKey:delegate.bannerRequest]) {
-        NSMutableOrderedSet *set = [[_bannerDelegates objectForKey:delegate.bannerRequest] mutableCopy];
-        if (![set containsObject:delegate]) {
-          GADMAdapterVungleMapTableRemoveObjectForKey(_bannerDelegates, delegate.bannerRequest);
-          [set addObject:delegate];
-          GADMAdapterVungleMapTableSetObjectForKey(_bannerDelegates, delegate.bannerRequest, set);
-          _bannerPlacementID = [delegate.desiredPlacement copy];
-        }
+        GADMAdapterVungleMapTableSetObjectForKey(_bannerDelegates, delegate.bannerRequest, delegate);
+        _bannerRequest = [delegate.bannerRequest copy];
+      } else {
+        id<GADMAdapterVungleDelegate> bannerDelegate = [_bannerDelegates objectForKey:delegate.bannerRequest];
+
+        /* The isRequestingBannerAdForRefresh flag is used for an edge case, when the old Banner
+         * delegate is removed from _bannerDelegates and there is a refresh Banner delegate doesn't
+         * construct Banner view successfully and add to _bannerDelegates yet. Adapter cannot set
+         * _bannerRequest to nil to prevent requesting another Banner ad with different bannerRequest.
+         */
+        bannerDelegate.isRequestingBannerAdForRefresh = YES;
```

Comment:
It took me a while to understand why you had these two booleans but I get it now. However, I think you could accomplish the same thing without those 2 boolean flags:

In this method, you could overwrite the NSMapTable with the new delegate after confirming it's the same banner request.

Then, in the removeDelegate method, you could write:

    if (delegate && (delegate == [_bannerDelegates objectForKey:delegate.bannerRequest])) {
      return;
    }
    GADMAdapterVungleMapTableRemoveObjectForKey(_bannerDelegates, delegate.bannerRequest);

Since the refreshed ad will already have taken over in the NSMapTable, when the old delegate tries to remove itself, it wouldn't pass the `delegate == [_bannerDelegates objectForKey:delegate.bannerRequest])` check. And it wouldn't get removed. Additionally, _bannerRequest stays intact the entire time.

Reply:
@ericleich Hi Eric, if we overwrite the NSMapTable with the new delegate after confirming it's the same banner request before the new refresh banner view is constructed successfully, this will cause problems in our next version which supports new event handlers, such as click event. If users click the old banner view, then adapter cannot send the click event to the old delegate which is replaced by the new refresh banner delegate. Thanks a lot. 🙂

Reply:
Ok, let's keep this then. In theory, it shouldn't matter because this should be an ad refresh, so the old instance is going away anyways. However, if Vungle ad fails to load and ultimately the request is a no fill, the old ad will still be there. So better to not remove the delegate.